src/DFG.cpp

```diff
@@ -8,1798 +8,1971 @@
-}
-
-// extract DFG from specific function
-void DFG::construct(Function& t_F) {
-
-  m_DFGEdges.clear();
-  nodes.clear();
-  m_ctrlEdges.clear();
-
-  int nodeID = 0;
-  int ctrlEdgeID = 0;
-  int dfgEdgeID = 0;
-
-  cout<<"*** current function: "<<t_F.getName().str()<<"\n";
-
-  // FIXME: eleminate duplicated edges.
-  for (Function::iterator BB=t_F.begin(), BEnd=t_F.end();
-      BB!=BEnd; ++BB) {
-    BasicBlock *curBB = &*BB;
-    errs()<<"*** current basic block: "<<*curBB->begin()<<"\n";
-    for (BasicBlock* sucBB : successors(curBB)) {
-      errs()<<"   ****** succ bb: "<<*sucBB->begin()<<"\n";
-    }
-
+ #include <fstream>
+ #include "DFG.h"
+ 
+ DFG::DFG(Function& t_F, list<Loop*>* t_loops, bool t_targetFunction,
+          bool t_precisionAware, bool t_heterogeneity,
+          map<string, int>* t_execLatency, list<string>* t_pipelinedOpt,
+    bool t_supportDVFS, bool t_DVFSAwareMapping,
+    int t_vectorFactorForIdiv) {
+   m_num = 0;
+   m_targetFunction = t_targetFunction;
+   m_targetLoops = t_loops;
+   m_orderedNodes = NULL;
+   m_CDFGFused = false;
+   m_cycleNodeLists = new list<list<DFGNode*>*>();
+   m_precisionAware = t_precisionAware;
+   m_supportDVFS = t_supportDVFS;
+   m_DVFSAwareMapping = t_DVFSAwareMapping;
+   m_vectorFactorForIdiv = t_vectorFactorForIdiv;
+ 
+   construct(t_F);
+ //  tuneForBranch();
+ //  tuneForLoad();
+   if (t_heterogeneity) { // parameter t_hetrogeneity is deprecated
+     calculateCycles();
+     nonlinear_combine();      // fusion for nonlinear ops
+ //    calculateCycles();
+ //    tuneForPattern();
+   }
+ //  trimForStandalone();
+   initExecLatency(t_execLatency);
+   initPipelinedOpt(t_pipelinedOpt);
+ 
+   // Pre-assigns the DVFS levels to each DFG node.
+   // This needs to be done after construct function
+   // as we need assign the highest frequency to the
+   // nodes on the critical path in the DFG.
+ }
+ 
+ void DFG::initDVFSLatencyMultiple(int t_II, int t_DVFSIslandDim,
+                       int t_numTiles) {
+   list<list<DFGNode*>*>* cycles = getCycleLists();
+   float max_cycle_length = 1.0;
+   for (list<DFGNode*>* cycle: *cycles) {
+     if (cycle->size() > max_cycle_length) {
+       max_cycle_length = cycle->size();
+     }
+   }
+   set<DFGNode*> assigned_dvfs_nodes;
+   int high_dvfs_dfg_nodes = 0;
+   int mid_dvfs_dfg_nodes = 0;
+   int low_dvfs_dfg_nodes = 0;
+   // TODO: might need to assign DVFS level based on the
+   // number of available CGRA nodes/resources.
+   for (list<DFGNode*>* cycle: *cycles) {
+     if (cycle->size() > max_cycle_length / 2) {
+       for (auto dfg_node : *cycle) {
+         dfg_node->setDVFSLatencyMultiple(1);
+   assigned_dvfs_nodes.insert(dfg_node);
+   high_dvfs_dfg_nodes += 1;
+       }
+     } else {
+       for (auto dfg_node : *cycle) {
+   if (assigned_dvfs_nodes.count(dfg_node) == 0) {
+           dfg_node->setDVFSLatencyMultiple(2);
+     assigned_dvfs_nodes.insert(dfg_node);
+     mid_dvfs_dfg_nodes += 1;
+   }
```

Comment:
The format (indent) seems messed up. Can you make a cleanup PR for the fix? Let's use "two space" rather than "tab".